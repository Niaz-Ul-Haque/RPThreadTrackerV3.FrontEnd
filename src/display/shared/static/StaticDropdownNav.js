import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
	activeTab: PropTypes.string.isRequired,
	setActiveTab: PropTypes.func.isRequired,
	options: PropTypes.arrayOf(PropTypes.shape({})).isRequired
};
const StaticDropdownNav = (props) => {
	const { activeTab, setActiveTab, options } = props;
	const optionElements = options.map(o => (
		<option value={o.tabId} selected={o.tabId === activeTab}>{o.name}</option>
	));
	return (
		<select className="clean-select" onChange={e => setActiveTab(e.target.value)}>
			{optionElements}
		</select>
	);
};
StaticDropdownNav.propTypes = propTypes;
export default StaticDropdownNav;
