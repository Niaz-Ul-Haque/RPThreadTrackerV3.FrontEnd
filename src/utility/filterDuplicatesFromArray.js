const filterDuplicatesFromArray = (array) => {
	return array.filter((elem, pos, arr) => elem && arr.indexOf(elem) == pos);
}
export default filterDuplicatesFromArray;
