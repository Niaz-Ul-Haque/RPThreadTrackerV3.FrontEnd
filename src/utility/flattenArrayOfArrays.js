const flattenArrayOfArrays = array => array.reduce((a, b) => a.concat(
	Array.isArray(b) ? flattenArrayOfArrays(b) : b
), []);
export default flattenArrayOfArrays;
