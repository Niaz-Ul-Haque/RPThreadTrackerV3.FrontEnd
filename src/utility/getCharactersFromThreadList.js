import { filterDuplicatesFromArray } from './';

const getCharactersFromThreadList = (threads) => {
	if (!threads || threads.length < 1) {
		return [];
	}
	const characters = threads.map(t => t.thread.character);
	return filterDuplicatesFromArray(characters, 'characterId');
};
export default getCharactersFromThreadList;
