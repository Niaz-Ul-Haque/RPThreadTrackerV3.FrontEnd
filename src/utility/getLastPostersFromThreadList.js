import { filterDuplicatesFromArray } from './';

const getLastPostersFromThreadList = (threads) => {
	if (!threads || threads.length < 1) {
		return [];
	}
	const lastPosters = threads.map((t) => {
		if (!t.status) {
			return '';
		}
		return t.status.LastPosterUrlIdentifier;
	});
	return filterDuplicatesFromArray(lastPosters);
};
export default getLastPostersFromThreadList;
