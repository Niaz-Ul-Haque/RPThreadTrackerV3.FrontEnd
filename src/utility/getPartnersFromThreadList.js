import { filterDuplicatesFromArray } from './';

const getPartnersFromThreadList = (threads) => {
	if (!threads || threads.length < 1) {
		return [];
	}
	const partners = threads.map((t) => {
		if (!t.thread) {
			return '';
		}
		return t.thread.partnerUrlIdentifier;
	});
	return filterDuplicatesFromArray(partners);
};
export default getPartnersFromThreadList;
