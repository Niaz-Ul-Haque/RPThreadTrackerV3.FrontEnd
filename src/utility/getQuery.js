import queryString from 'query-string';

const getQuery = () => {
	try {
		if (typeof (window) !== 'undefined') {
			return queryString.parse(window.location.search);
		}
		return {};
	} catch (e) {
		console.log(e);
		return null;
	}
};
export default getQuery;
