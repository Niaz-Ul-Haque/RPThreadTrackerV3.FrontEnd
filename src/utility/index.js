export { default as getQuery } from './getQuery';
export { default as flattenArrayOfArrays } from './flattenArrayOfArrays';
export { default as filterDuplicatesFromArray } from './filterDuplicatesFromArray';
export { default as getCharactersFromThreadList } from './getCharactersFromThreadList';
export { default as getPartnersFromThreadList } from './getPartnersFromThreadList';
export { default as getLastPostersFromThreadList } from './getLastPostersFromThreadList';
export { default as getTagsFromThreadList } from './getTagsFromThreadList';
