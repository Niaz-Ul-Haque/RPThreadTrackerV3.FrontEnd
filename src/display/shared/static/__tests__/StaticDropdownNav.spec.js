// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '~/testhelpers/helpers.unit';
import * as history from '../../../../utility/history';
import StaticDropdownNav from '../StaticDropdownNav';
// #endregion imports

jest.mock('../../../../utility/history', () => ({
	navigation: {
		navigateTo: jest.fn()
	}
}));

const createTestProps = (propOverrides) => ({
	activeTab: '/test/mock-option-2',
	options: [
		{
			href: '/test/mock-option-1',
			name: 'Mock Option 1'
		},
		{
			href: '/test/mock-option-2',
			name: 'Mock Option 2'
		}
	],
	...propOverrides
});
describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const jsx = <StaticDropdownNav {...props} />;
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
	describe('content', () => {
		it('should display an option for every option in props', () => {
			const props = createTestProps();
			const jsx = <StaticDropdownNav {...props} />;
			const element = shallow(jsx);
			const options = getSpecWrapper(element, 'static-dropdown-nav-option');
			expect(options).toHaveLength(2);
		});
	});
});

describe('behavior', () => {
	describe('setActiveTab', () => {
		it('should trigger navigation with selected value on change', () => {
			const mockEvent = {
				target: {
					value: '/test/mock-option-1'
				}
			};
			const props = createTestProps();
			const jsx = <StaticDropdownNav {...props} />;
			const element = shallow(jsx);
			const field = getSpecWrapper(element, 'static-dropdown-nav-select');
			field.simulate('change', mockEvent);
			expect(history.navigation.navigateTo).toHaveBeenCalledTimes(1);
			expect(history.navigation.navigateTo).toHaveBeenLastCalledWith('/test/mock-option-1');
		});
	});
});
