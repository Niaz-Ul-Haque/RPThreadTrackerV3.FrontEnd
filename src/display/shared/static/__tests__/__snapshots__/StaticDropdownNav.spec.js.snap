// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<styled.span>
  <select
    className="clean-select"
    data-spec="static-dropdown-nav-select"
    onChange={[Function]}
    value="test-tab"
  >
    <option
      data-spec="static-dropdown-nav-option"
      key="mock-option-1"
      value="mock-option-1"
    >
      Mock Option 1
    </option>
    <option
      data-spec="static-dropdown-nav-option"
      key="mock-option-2"
      value="mock-option-2"
    >
      Mock Option 2
    </option>
  </select>
</styled.span>
`;
